### ISSUE TITLE
Nvidia-docker containers fail to initialize with a CUDA error: out of memory

### ISSUE THREAD

**Issue by dtrash1** [NVIDIA/nvidia-docker#831]
### 1. Issue or feature description
Nvidia-docker containers always fail to initialize with a CUDA error: out of memory.  The docker install seems to be OK in that it can run non-nvidia images successfully (i.e. the hello-world image works).
### 2. Steps to reproduce the issue
1. Install using the Quickstart section of README.md
2. Upon running the final (test) step:
```
$ docker run --runtime=nvidia --rm nvidia/cuda:9.0-base nvidia-smi
```
 Receive the error:
```
docker: Error response from daemon: OCI runtime create failed: container_linux.go:348: starting container process caused "process_linux.go:402: container init caused \"process_linux.go:385: running prestart hook 1 caused \\\"error running hook: exit status 1, stdout: , stderr: exec command: [/usr/bin/nvidia-container-cli --load-kmods configure --ldconfig=@/sbin/ldconfig.real --device=all --compute --utility --require=cuda>=9.0 --pid=6610 /var/lib/docker/overlay2/e13075e398b2d91aee86c6a26fab25cb44c09987b6b099f09e71f16746a6b53d/merged]\\\\nnvidia-container-cli: initialization error: cuda error: out of memory\\\\n\\\"\"": unknown.
```
I have 3 1080Ti cards and one NVS310 that drives the display.  setting `NVIDIA_VISIBLE_DEVICES=2` (for example) does change the `--device=` argument for `nvidia-container-cli` but the returned error is essentially identical.

I'm not really sure how to proceed with troubleshooting, and haven't found anyone else with a similar problem.

### 3. Information to [attach](https://help.github.com/articles/file-attachments-on-issues-and-pull-requests/) (optional if deemed irrelevant)

 - [x] Kernel version from `uname -a`
```
Linux inslindl2 4.15.0-34-generic #37-Ubuntu SMP Mon Aug 27 15:21:48 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux
```
 - [ ] Any relevant kernel output lines from `dmesg`
 - [x] Driver information from `nvidia-smi -a`
The basic info:
```
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|===============================+======================+======================|
|   0  NVS 310             Off  | 00000000:05:00.0 N/A |                  N/A |
| 30%   56C    P0    N/A /  N/A |    244MiB /   451MiB |     N/A      Default |
+-------------------------------+----------------------+----------------------+
|   1  GeForce GTX 108...  Off  | 00000000:06:00.0 Off |                  N/A |
| 28%   42C    P8    10W / 250W |      2MiB / 11178MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+
|   2  GeForce GTX 108...  Off  | 00000000:09:00.0 Off |                  N/A |
| 28%   39C    P8     9W / 250W |      2MiB / 11178MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+
|   3  GeForce GTX 108...  Off  | 00000000:0A:00.0 Off |                  N/A |
|  0%   41C    P8     8W / 250W |      2MiB / 11178MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+
                                                                               
+-----------------------------------------------------------------------------+
| Processes:                                                       GPU Memory |
|  GPU       PID   Type   Process name                             Usage      |
|=============================================================================|
|    0                    Not Supported                                       |
+-----------------------------------------------------------------------------+

```
And the whole output:
[nvidia-smi.out.txt](https://github.com/NVIDIA/nvidia-docker/files/2425943/nvidia-smi.out.txt)
 - [x] Docker version from `docker version`
```
Client:
 Version:           18.06.1-ce
 API version:       1.38
 Go version:        go1.10.3
 Git commit:        e68fc7a
 Built:             Tue Aug 21 17:24:51 2018
 OS/Arch:           linux/amd64
 Experimental:      false

Server:
 Engine:
  Version:          18.06.1-ce
  API version:      1.38 (minimum version 1.12)
  Go version:       go1.10.3
  Git commit:       e68fc7a
  Built:            Tue Aug 21 17:23:15 2018
  OS/Arch:          linux/amd64
  Experimental:     false
```
 - [x] NVIDIA packages version from `dpkg -l '*nvidia*'` _or_ `rpm -qa '*nvidia*'`
```
Desired=Unknown/Install/Remove/Purge/Hold
| Status=Not/Inst/Conf-files/Unpacked/halF-conf/Half-inst/trig-aWait/Trig-pend
|/ Err?=(none)/Reinst-required (Status,Err: uppercase=bad)
||/ Name           Version      Architecture Description
+++-==============-============-============-=================================
un  libgldispatch0 <none>       <none>       (no description available)
ii  libnvidia-cfg1 390.48-0ubun amd64        NVIDIA binary OpenGL/GLX configur
un  libnvidia-cfg1 <none>       <none>       (no description available)
un  libnvidia-comm <none>       <none>       (no description available)
ii  libnvidia-comm 390.48-0ubun all          Shared files used by the NVIDIA l
ii  libnvidia-comp 390.48-0ubun amd64        NVIDIA libcompute package
ii  libnvidia-comp 390.48-0ubun i386         NVIDIA libcompute package
ii  libnvidia-cont 1.0.0-1      amd64        NVIDIA container runtime library 
ii  libnvidia-cont 1.0.0-1      amd64        NVIDIA container runtime library
un  libnvidia-deco <none>       <none>       (no description available)
ii  libnvidia-deco 390.48-0ubun amd64        NVIDIA Video Decoding runtime lib
ii  libnvidia-deco 390.48-0ubun i386         NVIDIA Video Decoding runtime lib
un  libnvidia-enco <none>       <none>       (no description available)
ii  libnvidia-enco 390.48-0ubun amd64        NVENC Video Encoding runtime libr
ii  libnvidia-enco 390.48-0ubun i386         NVENC Video Encoding runtime libr
un  libnvidia-fbc1 <none>       <none>       (no description available)
ii  libnvidia-fbc1 390.48-0ubun amd64        NVIDIA OpenGL-based Framebuffer C
ii  libnvidia-fbc1 390.48-0ubun i386         NVIDIA OpenGL-based Framebuffer C
un  libnvidia-gl   <none>       <none>       (no description available)
ii  libnvidia-gl-3 390.48-0ubun amd64        NVIDIA OpenGL/GLX/EGL/GLES GLVND 
ii  libnvidia-gl-3 390.48-0ubun i386         NVIDIA OpenGL/GLX/EGL/GLES GLVND 
un  libnvidia-ifr1 <none>       <none>       (no description available)
ii  libnvidia-ifr1 390.48-0ubun amd64        NVIDIA OpenGL-based Inband Frame 
ii  libnvidia-ifr1 390.48-0ubun i386         NVIDIA OpenGL-based Inband Frame 
un  nvidia-304     <none>       <none>       (no description available)
un  nvidia-340     <none>       <none>       (no description available)
un  nvidia-384     <none>       <none>       (no description available)
un  nvidia-390     <none>       <none>       (no description available)
un  nvidia-common  <none>       <none>       (no description available)
ii  nvidia-compute 390.48-0ubun amd64        NVIDIA compute utilities
ii  nvidia-contain 2.0.0+docker amd64        NVIDIA container runtime
ii  nvidia-contain 1.4.0-1      amd64        NVIDIA container runtime hook
ii  nvidia-dkms-39 390.48-0ubun amd64        NVIDIA DKMS package
un  nvidia-dkms-ke <none>       <none>       (no description available)
un  nvidia-docker  <none>       <none>       (no description available)
ii  nvidia-docker2 2.0.3+docker all          nvidia-docker CLI wrapper
ii  nvidia-driver- 390.48-0ubun amd64        NVIDIA driver metapackage
un  nvidia-driver- <none>       <none>       (no description available)
un  nvidia-kernel- <none>       <none>       (no description available)
ii  nvidia-kernel- 390.48-0ubun amd64        Shared files used with the kernel
un  nvidia-kernel- <none>       <none>       (no description available)
ii  nvidia-kernel- 390.48-0ubun amd64        NVIDIA kernel source package
un  nvidia-legacy- <none>       <none>       (no description available)
un  nvidia-libopen <none>       <none>       (no description available)
un  nvidia-opencl- <none>       <none>       (no description available)
un  nvidia-persist <none>       <none>       (no description available)
ii  nvidia-prime   0.8.8        all          Tools to enable NVIDIA's Prime
ii  nvidia-setting 390.42-0ubun amd64        Tool for configuring the NVIDIA g
un  nvidia-setting <none>       <none>       (no description available)
un  nvidia-smi     <none>       <none>       (no description available)
un  nvidia-utils   <none>       <none>       (no description available)
ii  nvidia-utils-3 390.48-0ubun amd64        NVIDIA driver support binaries
un  nvidia-vdpau-d <none>       <none>       (no description available)
ii  xserver-xorg-v 390.48-0ubun amd64        NVIDIA binary Xorg driver

```
 - [x] NVIDIA container library version from `nvidia-container-cli -V`
```
version: 1.0.0
build date: 2018-09-20T20:19+00:00
build revision: 881c88e2e5bb682c9bb14e68bd165cfb64563bb1
build compiler: x86_64-linux-gnu-gcc-7 7.3.0
build platform: x86_64
build flags: -D_GNU_SOURCE -D_FORTIFY_SOURCE=2 -DNDEBUG -std=gnu11 -O2 -g -fdata-sections -ffunction-sections -fstack-protector -fno-strict-aliasing -fvisibility=hidden -Wall -Wextra -Wcast-align -Wpointer-arith -Wmissing-prototypes -Wnonnull -Wwrite-strings -Wlogical-op -Wformat=2 -Wmissing-format-attribute -Winit-self -Wshadow -Wstrict-prototypes -Wunreachable-code -Wconversion -Wsign-conversion -Wno-unknown-warning-option -Wno-format-extra-args -Wno-gnu-alignof-expression -Wl,-zrelro -Wl,-znow -Wl,-zdefs -Wl,--gc-sections
```
 - [x] NVIDIA container library logs (see [troubleshooting](https://github.com/NVIDIA/nvidia-docker/wiki/Troubleshooting))
From `/var/log/nvidia-container-runtime-hook.log`
```
I0927 21:25:32.311741 7298 nvc.c:281] initializing library context (version=1.0.0, build=881c88e2e5bb682c9bb14e68bd165cfb64563bb1)
I0927 21:25:32.311821 7298 nvc.c:255] using root /
I0927 21:25:32.311833 7298 nvc.c:256] using ldcache /etc/ld.so.cache
I0927 21:25:32.311843 7298 nvc.c:257] using unprivileged user 65534:65534
I0927 21:25:32.314196 7304 nvc.c:191] loading kernel module nvidia
I0927 21:25:32.314496 7304 nvc.c:203] loading kernel module nvidia_uvm
I0927 21:25:32.314652 7304 nvc.c:211] loading kernel module nvidia_modeset
I0927 21:25:32.314998 7305 driver.c:133] starting driver service
I0927 21:25:32.403102 7298 driver.c:233] driver service terminated with signal 15
```
 - [x] Docker command, image and tag used
```
$ docker run --runtime=nvidia --env NVIDIA_VISIBLE_DEVICES=2 --rm nvidia/cuda:9.0-base bash
```

**Comment by flx42** [comment]
Before honoring `NVIDIA_VISIBLE_DEVICES`, we need to enumerate to initialize a CUDA context and enumerate the devices. Here it fails since your first GPU is old and doesn't have a lot of memory (and some of the memory is already taken, for the display probably).
We don't have a way to exclude GPUs here, so for now it's probably simpler if you remove this first GPU.

**Comment by 3XX0** [comment]
@dtrash1 you can exclude your NVS card see if it works:

```
sudo tee /etc/systemd/system/docker.service.d/override.conf <<EOF
[Service]
ExecStart=
ExecStart=/usr/bin/dockerd --host=fd://
EOF

nvidia-container-cli list -d 1,2,3 | sed 's/.*/DeviceAllow=& rwm/' | sudo tee -a /etc/systemd/system/docker.service.d/override.conf

sudo systemctl restart docker
```

**Comment by dtrash1** [comment]
Thanks so much for the help.  Excluding the NVS card as suggested bu @3XX0 allows me to run `nvidia-smi` inside the container and see the three 1080Ti's as available.

I do find it strange that I have no problem creating CUDA contexts and even running a locally installed (not containerized) version of Tensorflow without any precautions surrounding the NVS on this hardware.  I am running an older CUDA (8.0) with that config, however.

**Comment by 3XX0** [comment]
@dtrash1 Do you mean you don't get the above issue when running non-containerized? Did you try a CUDA sample and are you sure Tensorflow doesn't just ignore the NVS card?

**Comment by dtrash1** [comment]
@3XX0 That's correct.  I'd never seen this problem before, even though I've been using CUDA and Tensorflow for about a year now.  When I first installed CUDA, I built and ran the included samples/tests and (almost) everything worked perfectly.  It was certainly getting past the context create stage.  I am also sure that Tensorflow doesn't silently ignore the NVS, as I've mistakenly targeted it for a training run.  That does promptly fail with an "out of memory" error as Tensorflow tries to allocate space for the network layers.
